### PR TITLE
Streaming read optimization.

### PR DESCRIPTION
**Pull request by clin88**
This PR refactors deserialization to read directly from memory. That is, the serialization/deserialization path used to allocate intermediate representations called "Values" before doing a second pass that converted the values into primitive python objects. This process introduces considerable overhead however. By removing this layer and deserializing directly into python objects, we can dramatically improve performance.

Benchmarks:

Before:
```
--------------------------------------------------------------------- benchmark: 8 tests ---------------------------------------------------------------------
Name (time in us)                                      Min           Max          Mean      StdDev        Median          IQR  Outliers(*)  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------
test_binary_loads[nested_structs]              16,474.0086   33,906.9366   20,297.8651  5,972.9826   17,036.4380   2,691.3881        11;11      48           1
test_binary_loads[primitive_containers]       352,126.8368  374,081.8501  363,660.1448  8,221.5467  362,915.9927  10,945.7970          2;0       5           1
```

After:
```
--------------------------------------------------------------------- benchmark: 8 tests ---------------------------------------------------------------------
Name (time in us)                                      Min           Max          Mean      StdDev        Median          IQR  Outliers(*)  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------
test_binary_loads[nested_structs]              10,932.9224   29,603.0045   12,143.8092  2,779.9176   11,502.9812    415.9808          3;9      87           1
test_binary_loads[primitive_containers]       103,006.1245  111,021.9955  106,945.2498  2,231.4418  106,887.1021  2,384.5434          2;0       9           1
```

### CODE REVIEW

File: tests/util/spec.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from thriftrw.spec.set import SetTypeSpec
+from thriftrw.idl import Parser
+from thriftrw.compile.scope import Scope
+
+
+parse = Parser(start='struct', silent=True).parse
+
+
+def sstruct(fields):
+    struct_ast = parse('struct RefStruct {{ {} }}'.format(fields))
```
Comment (abhinav) [inline tests/util/spec.py]: Use `%s` please. This `{{ {} }}` thing is weird.

File: tests/util/spec.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+from thriftrw.spec.map import MapTypeSpec
+from thriftrw.spec.struct import StructTypeSpec
+from thriftrw.spec.union import UnionTypeSpec
+from thriftrw.spec.field import FieldSpec
+from thriftrw.spec.set import SetTypeSpec
+from thriftrw.idl import Parser
+from thriftrw.compile.scope import Scope
+
+
+parse = Parser(start='struct', silent=True).parse
```
Comment (abhinav) [inline tests/util/spec.py]: nit: `parse_struct`

File: tests/util/spec.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    spec.link(Scope("test"))
+    return spec
+
+
+def smap(key, value):
+    return MapTypeSpec(key, value)
+
+
+def sunion(*fields):
+    return UnionTypeSpec("test", list(fields))
```
Comment (abhinav) [inline tests/util/spec.py]: why is this so different from `sstruct`?

File: tests/util/spec.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def sf(id, spec, name=None, required=False, default_value=None):
```
Comment (abhinav) [inline tests/util/spec.py]: docs or examples for the weird ones like this, `sunion`, etc. would be nice.

File: tests/util/spec.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+from thriftrw.spec.field import FieldSpec
+from thriftrw.spec.set import SetTypeSpec
+from thriftrw.idl import Parser
+from thriftrw.compile.scope import Scope
+
+
+parse = Parser(start='struct', silent=True).parse
+
+
+def sstruct(fields):
```
Comment (abhinav) [inline tests/util/spec.py]: Default `fields=""` since you're passing that in in once place?

File: thriftrw/_buffer.pyx [hunk 6]
```diff
@@ -167,6 +167,23 @@ cdef class ReadBuffer(object):
 
         return result
 
+    cpdef void skip(self, int count) except *:
+        """Seek ``count`` bytes without performing any memory copying.
```
Comment (abhinav) [inline thriftrw/_buffer.pyx]: `s/Seek/Skip`

File: thriftrw/protocol/binary.pyx [hunk 7]
```diff
@@ -98,6 +99,306 @@ cdef class BinaryProtocolReader(object):
...
+        return self.reader.take(length)
+
+    cdef void skip_binary(self) except *:
+        cdef int32_t length = self._i32()
+        self.reader.skip(length)
+
+    cdef FieldHeader read_field_begin(self):
+        cdef int8_t field_type = self._byte()
+        if field_type == STRUCT_END:
+            return FieldHeader(-1, -1)
```
Comment (abhinav) [inline thriftrw/protocol/binary.pyx]: make sure to document this contract somewhere obvious like the docs for ProtocolReader.

File: thriftrw/spec/primitive.pyx [hunk 8]
```diff
@@ -139,6 +141,10 @@ cdef class _TextTypeSpec(_TextualTypeSpec):
     name = str('string')
     surface = unicode
 
+    cpdef object read_from(_TextTypeSpec self, ProtocolReader reader):
+        # TODO: Is this right?
+        return unicode(reader.read_binary())
```
Comment (abhinav) [inline thriftrw/spec/primitive.pyx]: return reader.read_binary().decode('utf-8')

File: thriftrw/spec/service.pyx [hunk 9]
```diff
@@ -174,6 +175,39 @@ cdef class FunctionResultSpec(UnionTypeSpec):
...
+        cdef FieldHeader header
+
+        header = reader.read_field_begin()
+
+        # We use -1 to signify struct end due to cython constraints.
+        while header.type != -1:
+            spec = self._index.get((header.id, header.type), None)
+
+            if spec is None:
+                raise UnknownExceptionError(
```
Comment (abhinav) [inline thriftrw/spec/service.pyx]: It's an unknown exception only if `header.id > 0`. If `header.id == 0`, a
value was returned for a `void` method and we can safely ignore it.

File: thriftrw/spec/service.pyx [hunk 10]
```diff
@@ -174,6 +175,39 @@ cdef class FunctionResultSpec(UnionTypeSpec):
...
+                    (
+                        '"%s" received an unrecognized exception. '
+                        'Make sure your Thrift IDL is up-to-date with '
+                        'what the remote host is using.'
+                    ) % self.name,
+                )
+
+            val = spec.spec.read_from(reader) or spec.default_value
```
Comment (abhinav) [inline thriftrw/spec/service.pyx]: If the field was present, you can't not read it. `read_from` must return a
value and it must be used. Defaults are applied automatically by `__init__` so
you don't need to do this.

Besides, FieldResultSpec can't have defaults anyway; it's a union.

File: thriftrw/spec/struct.pyx [hunk 11]
```diff
@@ -163,6 +168,30 @@ cdef class StructTypeSpec(TypeSpec):
...
+        cdef FieldHeader header = reader.read_field_begin()
+
+        while header.type != -1:
+            spec = self._index.get((header.id, header.type), None)
+
+            # Unrecognized field--possibly different version of struct definition.
+            if spec is None:
+                reader.skip(header.type)
+            else:
+                val = spec.spec.read_from(reader) or spec.default_value
```
Comment (abhinav) [inline thriftrw/spec/struct.pyx]: same as above; don't apply default value

File: thriftrw/spec/union.pyx [hunk 12]
```diff
@@ -149,6 +158,49 @@ cdef class UnionTypeSpec(TypeSpec):
...
+
+        # We use a 0 attribute to signify struct end due to cython constraints.
+        while header.type != -1:
+            spec = self._index.get((header.id, header.type), None)
+
+            # Unrecognized field--possibly different version of struct definition.
+            if spec is None:
+                reader.skip(header.type)
+            else:
+                val = spec.spec.read_from(reader) or spec.default_value
```
Comment (abhinav) [inline thriftrw/spec/union.pyx]: no default

File: .gitignore [hunk 13]
```diff
@@ -13,3 +13,4 @@ node_modules
 *.so
 .benchmarks/
 .coverage
+.idea/
```
Comment (breerly) [inline .gitignore]: Chen was here.

File: tests/util/spec.py (hunk 3 above)
Comment (clin88) [inline tests/util/spec.py]: Coz it's not actually used and this is broken. Will remove.

File: tests/util/spec.py (hunk 4 above)
Comment (clin88) [inline tests/util/spec.py]: Also redundant.

File: tests/util/spec.py (hunk 1 above)
Comment (abhinav) [inline tests/util/spec.py]: I mean it's weirder than,

    'struct RefStruct { %s }' % fields

**Comment by breerly** [comment]
@blampe this is definitely a minor, 1.1.0. We're changing how functionality works in order to provide performance improvements, not fixing existing functionality.

Users should pin to `>= 1, <2` in their `requirements.in`, so that regenerating `requirements.txt` results in the latest 1.x release being installed.

Let's not use patches as a means to ship new functionality, only for fixing existing functionality.

File: .travis.yml [hunk 14]
```diff
@@ -7,7 +7,6 @@ env:
     - TOX_ENV=py33
     - TOX_ENV=py34
     - TOX_ENV=pypy
-    - TOX_ENV=pypy3
```
Comment (clin88) [inline .travis.yml]: Didn't—it's broken on master as well. Pip 9 doesn't support this version of pypy. ABG and I decided to drop support as there appears to be no way to configure a custom version of pip in tox. When pypy3 catches up you can bring it back.

**Comment by breerly** [comment]
@clin88 this is looking pretty good; can you update the CHANGELOG to mention the performance improvement and that pypy3 support is being dropped?

File: tests/protocol/test_binary.py [hunk 15]
```diff
@@ -25,136 +25,172 @@
...
+
 
 def reader_writer_ids(x):
     if isinstance(x, (list, value.Value)):
         return None
     return ttype.name_of(x)
 
 
-@pytest.mark.parametrize('typ, bs, value', [
+@pytest.mark.parametrize('typ, bs, spec, value', [
```
Comment (breerly) [inline tests/protocol/test_binary.py]: thank you for expanding on these tests

File: tests/util/spec.py [hunk 16]
```diff
@@ -0,0 +1,29 @@
+from thriftrw.spec.list import ListTypeSpec
+from thriftrw.spec.map import MapTypeSpec
+from thriftrw.spec.struct import StructTypeSpec
+from thriftrw.spec.set import SetTypeSpec
+from thriftrw.idl import Parser
+from thriftrw.compile.scope import Scope
```
Comment (willhug) [inline tests/util/spec.py]: ordering

**Pull request merged**